onnxruntime/core/optimizer/graph_transformer_factory.h

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/optimizer/graph_transformer.h"
+#include "core/optimizer/conv_activation_fusion.h"
+#include "core/optimizer/conv_add_fusion.h"
+#include "core/optimizer/identity_elimination.h"
+#include "core/optimizer/slice_elimination.h"
+#include "core/optimizer/unsqueeze_elimination.h"
+#include "core/optimizer/conv_bn_fusion.h"
+#include "core/optimizer/conv_mul_fusion.h"
+#include "core/optimizer/gemm_activation_fusion.h"
+#include "core/optimizer/matmul_add_fusion.h"
+#include "core/optimizer/insert_cast_transformer.h"
+
+using namespace ::onnxruntime::common;
+
+namespace onnxruntime {
+/*
+* Owns all the graph transformers.
+*/
+class GraphTransformerFactory {
+ public:
+  explicit GraphTransformerFactory() {
+    InitFactory();
+  }
+
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(GraphTransformerFactory);
+
+  std::vector<GraphTransformer*> GetTransformers(TransformerLevel level) const {
+    std::vector<GraphTransformer*> transformers;
```

Comment:
Why don't we just have a std::vector<const GraphTransformer*> member for each level and return a const reference to that here so we're not creating new vectors each time?

Reply:
We need to design for 2 scenarios
1. Given a level run all the transformers in which case this method will be called to fetch all transformers applicable to that level

2. Given a list of id run all the transformers

I initially had map level -> list of transformers but with this it was difficult to achieve scenario 2, because then I was searching for a transformer in all levels

Reply:
isn't level -> { list of transformer names } -> iterate and lookup in transformer_map_ less efficient than
level -> { pointers to transformers }

transformer_map_ would still own the transformers, but the per level lookup can just have raw pointers.

---
In reply to: [263109365](https://github.com/Microsoft/onnxruntime/pull/556#discussion_r263109365) [](ancestors = 263109365)